onnxruntime/core/providers/cpu/tensor/upsample.cc

```diff
@@ -334,20 +390,180 @@ void upsampleBilinear(
+  coeffs[1] = static_cast<float>(((cubic_coeff_a + 2) * abs_s - (cubic_coeff_a + 3)) * abs_s * abs_s + 1);
+  coeffs[2] = static_cast<float>(((cubic_coeff_a + 2) * (1 - abs_s) - (cubic_coeff_a + 3)) * (1 - abs_s) * (1 - abs_s) + 1);
+  coeffs[3] = static_cast<float>(((cubic_coeff_a * (2 - abs_s) - 5 * cubic_coeff_a) * (2 - abs_s) + 8 * cubic_coeff_a) * (2 - abs_s) - 4 * cubic_coeff_a);
+  return coeffs;
+}
+
+// Get the tensor data at the requested coordinate
+template <typename T>
+T GetDataForCoordinate(const T* Xdata,
+                       int64_t x, int64_t y,
+                       int64_t input_height, int64_t input_width) {
+  x = std::max(static_cast<int64_t>(0), std::min(x, input_width - 1));
+  y = std::max(static_cast<int64_t>(0), std::min(y, input_height - 1));
+  return Xdata[y * input_width + x];
+}
+
+// Computes cubic convolution interpolation in 1D
+template <typename T>
+float CubicInterpolation1D(const T* Xdata,
+                           int64_t x,
+                           int64_t y,
+                           int64_t input_height,
+                           int64_t input_width,
+                           std::array<float, CubicModeGridLength>& coeff_array,
+                           std::unordered_map<int64_t, float>& cache) {
+  // When calculating cubic interpolation we move the 4*4 grid across the original data and therefore there is
+  // opportunity to cache the results for previously seen combinations.
+  // Check if the result is already available in the cache
+  auto grid_start_pos = (y)*input_width + (x - 1);
+  if (cache.find(grid_start_pos) != cache.end()) {
+    return cache[grid_start_pos];
+  }
+
+  // get the neighbors in 1D and find interpolation for this dimension
+  // for 1D cubic interpolation 4 samples are used. 2 on the left and 2 on the right of x
+  float result = 0;
+  for (int i = 0, j = -1; i < static_cast<int>(CubicModeGridLength); i++, j++) {
+    auto orig_data = GetDataForCoordinate(Xdata, x + j, y, input_height, input_width);
+    result += static_cast<float>(coeff_array[i] * orig_data);
+  }
+  cache[grid_start_pos] = result;
+
+  return result;
+}
+
+template <typename T>
+void ResizeBiCubic(
+    int64_t batch_size,
+    int64_t num_channels,
+    int64_t input_height,
+    int64_t input_width,
+    float height_scale,
+    float width_scale,
+    float cubic_coeff_a,
+    bool use_extrapolation,
+    float extrapolation_value,
+    const std::vector<float>& roi,
+    const T* Xdata,
+    T* Ydata,
+    GetOriginalCoordinateFunc get_original_coordinate) {
+  auto output_width = static_cast<int64_t>(input_width * width_scale);
+  auto output_height = static_cast<int64_t>(input_height * height_scale);
+
+  std::vector<float> y_original;
+  std::vector<float> x_original;
+  std::unordered_map<float, std::array<float, CubicModeGridLength>> cubic_coeffs;
```

Comment:
>cubic_coeffs [](start = 68, length = 12)

emm, i think 2 vector is better here. One for each dimension and use the output_x/y as the index.